### PR TITLE
Fix MemoryManager ctor and use internal span ctor to improve performance

### PR DESCRIPTION
**Pull request by ahsonkhan**
- Check for manager == null in the constructor
- Add and use Span.DangerousCreate in the Memory.Span property getter (to remove unnecessary bounds checks).

Related corefx PR with performance and unit tests: https://github.com/dotnet/corefx/pull/28880

cc @dotnet/corefxlab-contrib, @jkotas, @pakrym, @davidfowl, @stephentoub

### CODE REVIEW

File: src/mscorlib/shared/System/Memory.cs [hunk 1]
```diff
@@ -294,7 +292,7 @@ public Span<T> Span
                 }
                 else if (_object != null)
                 {
-                    return new Span<T>((T[])_object, _index, _length & RemoveFlagsBitMask);
+                    return Span<T>.DangerousCreate(Unsafe.As<T[]>(_object), _index, _length & RemoveFlagsBitMask);
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Memory.cs]: I don't think we should do this. Previously this check provided a level of safety against a Memory being torn, where at least validation would be performed when getting out the span to ensure that the created span was still within the bounds of the array.  By removing those checks, races on reading/writing a memory in some shared field could result in getting out a span that points to unexpected places.

File: src/mscorlib/shared/System/ReadOnlyMemory.cs [hunk 2]
```diff
@@ -200,7 +201,7 @@ public ReadOnlySpan<T> Span
                 }
                 else if (_object != null)
                 {
-                    return new ReadOnlySpan<T>((T[])_object, _index, _length & RemoveFlagsBitMask);
+                    return ReadOnlySpan<T>.DangerousCreate(Unsafe.As<T[]>(_object), _index, _length & RemoveFlagsBitMask);
```
Comment (stephentoub) [inline src/mscorlib/shared/System/ReadOnlyMemory.cs]: Same concern as earlier.

File: src/mscorlib/shared/System/Memory.cs (hunk 1 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Memory.cs]: Agree with @stephentoub . I do not think we should do this.

File: src/mscorlib/shared/System/Memory.cs [hunk 3]
```diff
@@ -120,20 +120,17 @@ public Memory(T[] array, int start, int length)
...
+        /// </exception>
         /// <exception cref="System.ArgumentOutOfRangeException">
         /// Thrown when the specified <paramref name="start"/> or end index is not in the range (&lt;0 or &gt;=Length).
         /// </exception>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public Memory(MemoryManager<T> manager, int start, int length)
         {
             if (manager == null)
-            {
-                if (start != 0 || length != 0)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Memory.cs]: This is a breaking change. It is ok only if this gets ported to 2.1. Once we ship this in 2.1, it needs to stay this way.

Comment (stephentoub) [reply src/mscorlib/shared/System/Memory.cs]: My understanding is that everything in master until at least some time next week will end up in 2.1.

Comment (joshfree) [reply src/mscorlib/shared/System/Memory.cs]: Yes there will be a last master->release/2.1 integration on April 13 for RC.  After which the master branch becomes "2.2"

File: src/mscorlib/shared/System/Memory.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/mscorlib/shared/System/Memory.cs]: How do we assess the risk of struct tearing in such cases given there are other places, particularly in Memory, where we don't guard against this? We don't check if the object field and index/length are in sync while slicing a memory, and even when getting the span property for objects other than T[].

Since you guys think that at the point of accessing the span, we should keep this check, I will revert this.

Comment (jkotas) [reply src/mscorlib/shared/System/Memory.cs]: The design point of the framework has always been that threading bugs can get you into invalid state, but they should not cause type safety or memory safety violations.

**Comment by GrabYourPitchforks** [comment]
Why the concerns about struct tearing resulting in AVs for the `Span` property getter but not the `Pin()` method?

**Comment by stephentoub** [comment]
> Why the concerns about struct tearing resulting in AVs for the Span property getter but not the Pin() or Slice(...) methods?

Because I don't see the Pin method in this PR :wink:

If there's a similar issue there, it should be fixed.

**Comment by ahsonkhan** [comment]
> If there's a similar issue there, it should be fixed.

I will make the fix as part of this PR.

File: src/mscorlib/shared/System/Memory.cs [hunk 4]
```diff
@@ -276,6 +273,7 @@ public Span<T> Span
                 if (_index < 0)
                 {
                     Debug.Assert(_length >= 0);
+                    Debug.Assert(_object != null);
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Memory.cs]: The `if (_object != null)` case we discussed will be fixed in a separate PR?

Comment (stephentoub) [reply src/mscorlib/shared/System/Memory.cs]: Oh, wait, nevermind.

Comment (ahsonkhan) [reply src/mscorlib/shared/System/Memory.cs]: > The `if (_object != null)` case we discussed will be fixed in a separate PR?

Pin looks correct to me, so there was nothing to update.
